File: backend/Minimarket.Api/Services/CategoryService.cs
using Minimarket.Api.DTOs;
using Minimarket.Api.Mapping;
using Minimarket.Api.Models;
using Minimarket.Api.Repositories;

namespace Minimarket.Api.Services;

public class CategoryService(ICategoryRepository categoryRepository) : ICategoryService
{
    public async Task<IReadOnlyCollection<CategoryDto>> GetAllAsync(bool includeInactive) =>
        (await categoryRepository.GetAllAsync(includeInactive)).Select(x => x.ToDto()).ToList();

    public async Task<(bool Success, string? Error, CategoryDto? Category)> CreateAsync(SaveCategoryDto dto)
    {
        var normalizedName = dto.Name.Trim();
        if (string.IsNullOrWhiteSpace(normalizedName))
        {
            return (false, "El nombre de la categoria es obligatorio.", null);
        }

        if (await categoryRepository.GetByNameAsync(normalizedName) is not null)
        {
            return (false, "Ya existe una categoria con el mismo nombre.", null);
        }

        var category = new Category
        {
            Name = normalizedName,
            Description = dto.Description?.Trim(),
            IsActive = dto.IsActive
        };

        await categoryRepository.AddAsync(category);
        await categoryRepository.SaveChangesAsync();

        return (true, null, category.ToDto());
    }

    public async Task<(bool Success, string? Error, CategoryDto? Category)> UpdateAsync(int id, SaveCategoryDto dto)
    {
        var category = await categoryRepository.GetByIdAsync(id);
        if (category is null)
        {
            return (false, "Categoria no encontrada.", null);
        }

        var normalizedName = dto.Name.Trim();
        if (string.IsNullOrWhiteSpace(normalizedName))
        {
            return (false, "El nombre de la categoria es obligatorio.", null);
        }

        var byName = await categoryRepository.GetByNameAsync(normalizedName);
        if (byName is not null && byName.Id != id)
        {
            return (false, "Ya existe una categoria con el mismo nombre.", null);
        }

        // Si el nombre no cambio, byName es la misma fila ya rastreada; la reutilizamos para evitar
        // conflictos de seguimiento. Si cambio a un nombre nuevo, usamos la instancia cargada por Id.
        var toUpdate = byName is not null && byName.Id == id ? byName : category;
        toUpdate.Name = normalizedName;
        toUpdate.Description = dto.Description?.Trim();
        toUpdate.IsActive = dto.IsActive;

        categoryRepository.Update(toUpdate);
        await categoryRepository.SaveChangesAsync();

        return (true, null, toUpdate.ToDto());
    }
}

File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/service/CategoryService.java
package com.minimarket.api.service;

import com.minimarket.api.dto.CategoryDto;
import com.minimarket.api.dto.SaveCategoryDto;
import com.minimarket.api.entity.Category;
import com.minimarket.api.repository.CategoryRepository;
import com.minimarket.api.util.DtoMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

@Service
public class CategoryService {

    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    public List<CategoryDto> getAll(boolean includeInactive) {
        var categories = includeInactive
            ? categoryRepository.findAllByOrderByNameAsc()
            : categoryRepository.findByIsActiveTrueOrderByNameAsc();
        return categories.stream().map(DtoMapper::toDto).toList();
    }

    private BigDecimal normalizePriceAdjustment(BigDecimal value) {
        var normalized = value != null ? value : BigDecimal.ZERO;
        if (normalized.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("El ajuste de precio no puede ser negativo.");
        }

        return normalized.setScale(2, RoundingMode.HALF_UP);
    }

    @Transactional
    public ServiceResult<CategoryDto> create(SaveCategoryDto dto) {
        final BigDecimal normalizedAdjustment;
        try {
            normalizedAdjustment = normalizePriceAdjustment(dto.priceAdjustmentPercentage());
        } catch (IllegalArgumentException ex) {
            return ServiceResult.failure(ex.getMessage());
        }

        var normalizedName = dto.name() != null ? dto.name().trim() : "";
        if (normalizedName.isBlank()) {
            return ServiceResult.failure("El nombre de la categoria es obligatorio.");
        }

        if (categoryRepository.existsByNameIgnoreCase(normalizedName)) {
            return ServiceResult.failure("Ya existe una categoria con el mismo nombre.");
        }

        var category = new Category();
        category.setName(normalizedName);
        category.setDescription(dto.description() != null ? dto.description().trim() : null);
        category.setIsActive(Boolean.TRUE.equals(dto.isActive()));
        category.setPriceAdjustmentPercentage(normalizedAdjustment);

        var saved = categoryRepository.save(category);
        return ServiceResult.success(DtoMapper.toDto(saved));
    }

    @Transactional
    public ServiceResult<CategoryDto> update(Integer id, SaveCategoryDto dto) {
        final BigDecimal normalizedAdjustment;
        try {
            normalizedAdjustment = normalizePriceAdjustment(dto.priceAdjustmentPercentage());
        } catch (IllegalArgumentException ex) {
            return ServiceResult.failure(ex.getMessage());
        }

        var category = categoryRepository.findById(id).orElse(null);
        if (category == null) {
            return ServiceResult.failure("Categoria no encontrada.");
        }

        var normalizedName = dto.name() != null ? dto.name().trim() : "";
        if (normalizedName.isBlank()) {
            return ServiceResult.failure("El nombre de la categoria es obligatorio.");
        }

        var byName = categoryRepository.findByNameIgnoreCase(normalizedName).orElse(null);
        if (byName != null && !byName.getId().equals(id)) {
            return ServiceResult.failure("Ya existe una categoria con el mismo nombre.");
        }

        category.setName(normalizedName);
        category.setDescription(dto.description() != null ? dto.description().trim() : null);
        category.setIsActive(Boolean.TRUE.equals(dto.isActive()));
        category.setPriceAdjustmentPercentage(normalizedAdjustment);

        var saved = categoryRepository.save(category);
        return ServiceResult.success(DtoMapper.toDto(saved));
    }
}
